File: __tests__/Aeson_json_test.ml
open Jest
open Expect

let () = 

describe "json.t" (fun () ->
  let open Aeson in                        

  (* test unicode code points and regex *)
  test "test regex against unicode string" (fun () ->
    expect (Js.Re.test {js|"上"|js} [%re "/\\u4E0A/"]) |> toEqual true);

  test "test regex against unicode code point string" (fun () ->
    expect (Js.Re.test {js|"\u4E0A"|js} [%re "/\\u4E0A/"]) |> toEqual true);

  test "test regex against unicode code point string" (fun () ->
    expect (Js.Re.test (Js.String.fromCharCode 0x4E0A) [%re "/^\\u4E0A$/"]) |> toEqual true);  

  test "test regex against unicode code point string in private use area" (fun () ->
    expect (Js.Re.test (Js.String.fromCharCode 0xE000) [%re "/^\\uE000$/"]) |> toEqual true);  


  (* is_numeric *)
  test "is_numeric" (fun () ->
    expect (Js.Re.test ((Js.String.fromCharCode 0xE000) ^ "9.0") Json.is_numeric) |> toEqual true);
  
  
  (* parse *)
  test "dict" (fun () ->
    expect @@ (Json.stringify (Aeson.Json.parseExn {| { "a": "x", "b": "y" } |}))
      |> toEqual "{\"a\":\"x\",\"b\":\"y\"}"  );

  test "array" (fun () ->
    expect @@ (Json.stringify (Aeson.Json.parseExn {| [ "a", "x", "b", "y" ] |}))
      |> toEqual "[\"a\",\"x\",\"b\",\"y\"]"  );

  (* compare output from Aeson.Json.stringify and Js.Json.stringify *)
  test "dict" (fun () ->
    expect @@ (Json.stringify (Aeson.Json.parseExn {| { "a": "x", "b": "y" } |}))
      |> toEqual (Js.Json.stringify (Js.Json.parseExn {| { "a": "x", "b": "y" } |})));

  test "array" (fun () ->
    expect @@ (Json.stringify (Aeson.Json.parseExn {| [ "a", "x", "b", "y" ] |}))
      |> toEqual (Js.Json.stringify (Js.Json.parseExn {| [ "a", "x", "b", "y" ] |})));

  
  (* stringify *)
  test "int64" (fun () -> 
    expect (Json.stringify (Json.int64 (Int64.of_string "9223372036854775807"))) |> toEqual "9223372036854775807");

  test "bigint" (fun () -> 
    expect (Json.stringify (Json.bigint (Bigint.of_string "823891829388198398494893843948394892233720368547758072323123123"))) |> toEqual "823891829388198398494893843948394892233720368547758072323123123");

  test "string" (fun () ->
    expect @@ Json.stringify @@ Json.string "foo" |> toEqual "\"foo\"");

  (* decode *)
  
  
  (* encode *)


  (* Aeson.Json.t and Js.Json.t*)
  test "array" (fun () ->
    expect @@ (Json.to_js_json (Aeson.Json.parseExn {| [ "a", "x", "b", "y" ] |}))
      |> toEqual (Js.Json.parseExn {| [ "a", "x", "b", "y" ] |}));

  (* Aeson.Json.t and Js.Json.t*)
  test "array" (fun () ->
    expect @@ (Json.from_js_json (Js.Json.parseExn {| [ "a", "x", "b", "y" ] |}))
      |> toEqual (Aeson.Json.parseExn {| [ "a", "x", "b", "y" ] |}))

);

File: lib/js/__tests__/Aeson_json_test.js
'use strict';

var Jest = require("@glennsl/bs-jest/lib/js/src/jest.js");
var Bigint = require("bs-Zarith/lib/js/src/Bigint.js");
var Aeson_json = require("../src/Aeson_json.js");
var Caml_format = require("bs-platform/lib/js/caml_format.js");

describe("json.t", (function () {
        Jest.test("test regex against unicode string", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0]((/\u4E0A/).test("\"上\"")));
              }));
        Jest.test("test regex against unicode code point string", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0]((/\u4E0A/).test("\"\u4E0A\"")));
              }));
        Jest.test("test regex against unicode code point string", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0]((/^\u4E0A$/).test(String.fromCharCode(19978))));
              }));
        Jest.test("test regex against unicode code point string in private use area", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0]((/^\uE000$/).test(String.fromCharCode(57344))));
              }));
        Jest.test("is_numeric", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0](Aeson_json.is_numeric.test(String.fromCharCode(57344) + "9.0")));
              }));
        Jest.test("dict", (function () {
                return Jest.Expect[/* toEqual */12]("{\"a\":\"x\",\"b\":\"y\"}", Jest.Expect[/* expect */0](Aeson_json.stringify(JSON.parse(" { \"a\": \"x\", \"b\": \"y\" } "))));
              }));
        Jest.test("array", (function () {
                return Jest.Expect[/* toEqual */12]("[\"a\",\"x\",\"b\",\"y\"]", Jest.Expect[/* expect */0](Aeson_json.stringify(JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "))));
              }));
        Jest.test("dict", (function () {
                return Jest.Expect[/* toEqual */12](JSON.stringify(JSON.parse(" { \"a\": \"x\", \"b\": \"y\" } ")), Jest.Expect[/* expect */0](Aeson_json.stringify(JSON.parse(" { \"a\": \"x\", \"b\": \"y\" } "))));
              }));
        Jest.test("array", (function () {
                return Jest.Expect[/* toEqual */12](JSON.stringify(JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] ")), Jest.Expect[/* expect */0](Aeson_json.stringify(JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "))));
              }));
        Jest.test("int64", (function () {
                return Jest.Expect[/* toEqual */12]("9223372036854775807", Jest.Expect[/* expect */0](Aeson_json.stringify(Aeson_json.int64(Caml_format.caml_int64_of_string("9223372036854775807")))));
              }));
        Jest.test("bigint", (function () {
                return Jest.Expect[/* toEqual */12]("823891829388198398494893843948394892233720368547758072323123123", Jest.Expect[/* expect */0](Aeson_json.stringify(Aeson_json.bigint(Bigint.of_string("823891829388198398494893843948394892233720368547758072323123123")))));
              }));
        Jest.test("string", (function () {
                return Jest.Expect[/* toEqual */12]("\"foo\"", Jest.Expect[/* expect */0](Aeson_json.stringify("foo")));
              }));
        Jest.test("array", (function () {
                return Jest.Expect[/* toEqual */12](JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "), Jest.Expect[/* expect */0](Aeson_json.to_js_json(JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "))));
              }));
        return Jest.test("array", (function () {
                      return Jest.Expect[/* toEqual */12](JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "), Jest.Expect[/* expect */0](Aeson_json.from_js_json(JSON.parse(" [ \"a\", \"x\", \"b\", \"y\" ] "))));
                    }));
      }));

/*  Not a pure module */
